File: macos/Sources/ZhuoDaziMac/AnalyticsService.swift
import Foundation

final class AnalyticsService {
    private static let eventsURL = DeskPetApi.analyticsEvents
    private let installationId: String
    private let architecture: String
    private let firstLaunchMarkerURL: URL

    init(licenses: LicenseService) {
        installationId = licenses.installationId
#if arch(arm64)
        architecture = "arm64"
#else
        architecture = "x86_64"
#endif
        let directory = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("ZhuoDazi", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        firstLaunchMarkerURL = directory.appendingPathComponent("analytics-first-launch.marker")
    }

    func trackStartup() {
        let occurredAt = ISO8601DateFormatter().string(from: Date())
        let types = FileManager.default.fileExists(atPath: firstLaunchMarkerURL.path)
            ? ["app_session_start", "app_daily_active"]
            : ["app_first_launch", "app_session_start", "app_daily_active"]
        let events = types.map { type in
            [
                "eventId": "macos-\(UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased())",
                "type": type,
                "installationId": installationId,
                "platform": "macos",
                "architecture": architecture,
                "version": AppVersion.current,
                "occurredAt": occurredAt
            ]
        }
        var request = URLRequest(url: Self.eventsURL)
        request.httpMethod = "POST"
        request.timeoutInterval = 12
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("ZhuoDazi/\(AppVersion.current)", forHTTPHeaderField: "User-Agent")
        request.setValue("macos", forHTTPHeaderField: "X-DeskPet-Platform")
        request.setValue(architecture, forHTTPHeaderField: "X-DeskPet-Architecture")
        guard let body = try? JSONSerialization.data(withJSONObject: ["events": events]) else { return }
        request.httpBody = body
        Task {
            for attempt in 0..<3 {
                guard let (_, response) = try? await URLSession.shared.data(for: request),
                      let http = response as? HTTPURLResponse else {
                    if attempt < 2 { try? await Task.sleep(for: .seconds(attempt + 1)) }
                    continue
                }
                if (200..<300).contains(http.statusCode) {
                    if !FileManager.default.fileExists(atPath: firstLaunchMarkerURL.path) {
                        try? Data(occurredAt.utf8).write(to: firstLaunchMarkerURL, options: .atomic)
                    }
                    return
                }
                if attempt < 2 { try? await Task.sleep(for: .seconds(attempt + 1)) }
            }
        }
    }
}

File: windows/ZhuoDazi/Services/AnalyticsService.cs
using System.Net.Http;
using System.Text.Json;

namespace ZhuoDazi.Services;

public sealed class AnalyticsService : IDisposable
{
    private const string EventsUrl = DeskPetApi.AnalyticsEvents;
    private readonly LicenseService _licenses;
    private readonly HttpClient _httpClient = DeskPetHttp.CreateClient(TimeSpan.FromSeconds(12));
    private readonly string _firstLaunchMarkerPath;
    private bool _disposed;

    public AnalyticsService(LicenseService licenses)
    {
        _licenses = licenses;
        var dataDirectory = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "poko-desktop-pet");
        Directory.CreateDirectory(dataDirectory);
        _firstLaunchMarkerPath = Path.Combine(dataDirectory, "analytics-first-launch.marker");
    }

    public async Task TrackStartupAsync()
    {
        if (_disposed) return;
        var occurredAt = DateTimeOffset.UtcNow.ToString("O");
        var types = File.Exists(_firstLaunchMarkerPath)
            ? new[] { "app_session_start", "app_daily_active" }
            : new[] { "app_first_launch", "app_session_start", "app_daily_active" };
        var events = types.Select(type => CreateEvent(type, occurredAt)).ToArray();
        try
        {
            var payload = JsonSerializer.SerializeToUtf8Bytes(new { events });
            for (var attempt = 0; attempt < 3; attempt++)
            {
                using var request = new HttpRequestMessage(HttpMethod.Post, EventsUrl)
                {
                    Content = new ByteArrayContent(payload)
                };
                request.Content.Headers.ContentType = new("application/json") { CharSet = "utf-8" };
                request.Headers.UserAgent.ParseAdd($"ZhuoDazi/{UpdateService.CurrentVersion}");
                request.Headers.TryAddWithoutValidation("X-DeskPet-Platform", "windows");
                request.Headers.TryAddWithoutValidation("X-DeskPet-Architecture", "x64");
                using var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
                if (response.IsSuccessStatusCode)
                {
                    if (!File.Exists(_firstLaunchMarkerPath))
                        await File.WriteAllTextAsync(_firstLaunchMarkerPath, occurredAt);
                    return;
                }
                if (attempt < 2) await Task.Delay(TimeSpan.FromSeconds(attempt + 1));
            }
        }
        catch
        {
            // Metrics must never prevent the desktop pet from starting.
        }
    }

    private object CreateEvent(string type, string occurredAt) => new
    {
        eventId = $"windows-{Guid.NewGuid():N}",
        type,
        installationId = _licenses.InstallationId,
        platform = "windows",
        architecture = "x64",
        version = UpdateService.CurrentVersion,
        occurredAt
    };

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _httpClient.Dispose();
    }
}
